File: backend/src/main/java/com/devlaunch/service/interfaces/NotificationService.java
package com.devlaunch.service.interfaces;

import com.devlaunch.dto.response.NotificationResponse;
import com.devlaunch.entity.User;
import com.devlaunch.entity.enums.NotificationType;
import com.devlaunch.exception.ResourceNotFoundException;

import java.util.List;

/**
 * Service interface for the notification center.
 * <p>
 * Defines the contract for creating notifications automatically from the
 * existing platform modules, fanning out admin announcements to every
 * user, and managing the authenticated user's own notification inbox
 * (listing, unread counting, marking read, and deleting). All read and
 * write operations are scoped to the authenticated user so a user can
 * never access another user's notifications.
 * </p>
 *
 * @author DevLaunch
 */
public interface NotificationService {

    /**
     * Creates a notification for the given user.
     * <p>
     * Called internally by the existing modules (resume builder, job
     * tracker, study planner, AI features, admin announcements) whenever
     * an important action completes. New notifications always start unread.
     * </p>
     *
     * @param user    the user to notify
     * @param type    the notification category
     * @param title   the notification headline
     * @param message the notification body
     */
    void createNotification(User user, NotificationType type, String title, String message);

    /**
     * Creates an announcement notification for every registered user.
     * <p>
     * Used by the admin module when an announcement is published so all
     * users receive it as a notification without duplicating the
     * announcement storage itself.
     * </p>
     *
     * @param type    the notification category (typically {@code ANNOUNCEMENT})
     * @param title   the notification headline
     * @param message the notification body
     */
    void notifyAllUsers(NotificationType type, String title, String message);

    /**
     * Retrieves all notifications belonging to the currently authenticated
     * user, newest first.
     *
     * @return a list of the authenticated user's notifications,
     *         or an empty list if none exist
     */
    List<NotificationResponse> getNotifications();

    /**
     * Counts the unread notifications of the currently authenticated user.
     *
     * @return the number of unread notifications
     */
    long getUnreadCount();

    /**
     * Marks a single notification as read, verifying it belongs to the
     * currently authenticated user.
     *
     * @param id the notification ID to mark as read
     * @return the updated notification data
     * @throws ResourceNotFoundException if the notification is not found
     *                                   or does not belong to the user
     */
    NotificationResponse markAsRead(Long id);

    /**
     * Marks every notification of the currently authenticated user as read.
     *
     * @return the updated notification list, newest first
     */
    List<NotificationResponse> markAllAsRead();

    /**
     * Deletes a single notification, verifying it belongs to the currently
     * authenticated user.
     *
     * @param id the notification ID to delete
     * @throws ResourceNotFoundException if the notification is not found
     *                                   or does not belong to the user
     */
    void deleteNotification(Long id);

    /**
     * Deletes every notification belonging to the given user.
     * <p>
     * Used by the admin module when deleting a user so the
     * {@code notifications.user_id} foreign key constraint stays valid.
     * </p>
     *
     * @param user the user whose notifications to delete
     */
    void deleteAllForUser(User user);

}

File: frontend/src/services/notification.service.ts
/**
 * Notification service.
 *
 * Provides the notification center methods backed by the authenticated
 * /api/notifications endpoints. Every endpoint operates exclusively on
 * the currently logged-in user's own notifications.
 *
 * @see backend/src/main/java/com/devlaunch/controller/NotificationController.java
 * @author DevLaunch
 */

import apiClient from '../api/client';
import { NOTIFICATIONS } from '../api/endpoints';
import type { AppNotification } from '../types/notification';

export const notificationService = {
  /**
   * Retrieves all notifications for the authenticated user, newest first.
   *
   * GET /api/notifications
   */
  getAll: () =>
    apiClient.get<AppNotification[]>(NOTIFICATIONS.BASE).then((res) => res.data),

  /**
   * Retrieves the unread notification count for the authenticated user.
   *
   * GET /api/notifications/unread-count
   */
  getUnreadCount: () =>
    apiClient.get<number>(NOTIFICATIONS.UNREAD_COUNT).then((res) => res.data),

  /**
   * Marks a single notification as read.
   *
   * PUT /api/notifications/{id}/read
   */
  markAsRead: (id: number) =>
    apiClient.put<AppNotification>(NOTIFICATIONS.MARK_READ(id)).then((res) => res.data),

  /**
   * Marks every notification of the authenticated user as read.
   *
   * PUT /api/notifications/read-all
   */
  markAllAsRead: () =>
    apiClient.put<AppNotification[]>(NOTIFICATIONS.READ_ALL).then((res) => res.data),

  /**
   * Deletes a single notification.
   *
   * DELETE /api/notifications/{id}
   */
  deleteNotification: (id: number) =>
    apiClient.delete<void>(NOTIFICATIONS.BY_ID(id)).then(() => undefined),
};
